backend/src/main/java/com/javahelp/backend/query/UserQueryConstraint.java

```diff
@@ -0,0 +1,75 @@
+package com.javahelp.backend.query;
+
+import com.javahelp.model.user.User;
+import com.javahelp.backend.data.DynamoDBUserStore;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.io.*;
+
+/**
+ * This class subsets the list of Providers in the database based on query constraints
+ * specified by the user.
+ * The query constraints are indicated using filters (multiple choice) or searches (text name).
+ */
+public class UserQueryConstraint {
+    private DynamoDBUserStore dbUserStore;
+
+    public UserQueryConstraint(DynamoDBUserStore dbUserStore) {
+        this.dbUserStore = dbUserStore;
+    }
+
+    public HashSet<User> getProvidersWithMultiConstraints(List<Constraint> constraints) {
+        HashSet<User> providers = new HashSet<>();
+        HashSet<String> providersId;
+
+        for (int i = 0; i < constraints.size(); i++) {
+            providersId = IdSet(providers);
+
+            Constraint constraint = constraints.get(i);
+
+            HashSet<User> subsetProviders = getProvidersWithConstraint(constraint);
+            HashSet<String> subsetProvidersId = IdSet(subsetProviders);
+            if (providers.isEmpty()) {
+                providers = subsetProviders;
+            }
+            else {
+                providersId.retainAll(subsetProvidersId);
+                providers = getUserSetFromIdSet(providers, providersId);
+            }
```

Comment:
Instead of running multiple queries and combining the results, it is probably better to run a single query.

Reply:
Changed in recent push